tools/run_tests/run_xds_tests.py

```diff
@@ -1002,14 +1004,30 @@ def test_header_matching(gcp, original_backend_service, instance_group,
 
 
 def set_serving_status(instances, service_port, serving):
+    logger.info('setting %s serving status to %s', instances, serving)
     for instance in instances:
         with grpc.insecure_channel('%s:%d' %
                                    (instance, service_port)) as channel:
+            logger.info('setting %s serving status to %s', instance, serving)
             stub = test_pb2_grpc.XdsUpdateHealthServiceStub(channel)
-            if serving:
-                stub.SetServing(empty_pb2.Empty())
-            else:
-                stub.SetNotServing(empty_pb2.Empty())
+            health_stub = health_pb2_grpc.HealthStub(channel)
+
+            retry_count = 5
+            for i in range(5):
+                if serving:
+                    stub.SetServing(empty_pb2.Empty())
+                else:
+                    stub.SetNotServing(empty_pb2.Empty())
+                serving_status = health_stub.Check(
+                    health_pb2.HealthCheckRequest())
+                logger.info('got instance service status %s', serving_status)
+                want_status = health_pb2.HealthCheckResponse.SERVING if serving else health_pb2.HealthCheckResponse.NOT_SERVING
+                if serving_status.status == want_status:
+                    break
```

Comment:
Suggest changing this to `return`, then drop the `if` below and just raise the exception if the for loop terminates

Reply:
This cannot just return (in fact, I did that at first..), there's a outer for loop over all the instances.